docs/agents/sampling.md

```diff
@@ -0,0 +1,64 @@
+# Sampling
+
+To reduce processing and storage overhead, transactions may be sampled by agents.
+Sampling here refers to "head-based sampling".
+Head-based sampling is where a sampling decision is made at the root of the distributed trace, before the details or outcome of the trace are known, and propagated throughout the trace.
+
+## Configuration
+
+Agents can be configured to sample probabilistically, by specifying a sampling probability in the range \[0,1\] using the configuration `ELASTIC_APM_TRANSACTION_SAMPLE_RATE`. e.g.
+
+ - `ELASTIC_APM_TRANSACTION_SAMPLE_RATE=1` means all transactions will be sampled (the default)
+ - `ELASTIC_APM_TRANSACTION_SAMPLE_RATE=0` means no transactions will be sampled
+ - `ELASTIC_APM_TRANSACTION_SAMPLE_RATE=0.5` means approximately 50% of transactions will be sampled
+
+## Sampling rate
+
+At the time of making a sampling decision, the sampling rate must be recorded so that it can be associated with every transaction and span in the trace.
+This will be used by the server for scaling transaction and span metrics.
+
+The sampling rate must be recorded on transactions and spans as `sample_rate`. e.g.
+
+    {"transaction":{"name":"GET /","sample_rate":0.1,...}}
+    {"span":{"name":"SELECT FROM table","sample_rate":0.1,...}}
+
+For non-sampled transactions the `sample_rate` field _must_ be set to 0.
+For backwards compatibility the server will assume a value of 1 where the field is unspecified,
+resulting in all transactions (sampled and unsampled) being counted equally in metrics.
+
+## Non-sampled transactions
+
+Currently _all_ transactions are captured by Elastic APM agents.
+Sampling controls how much data is captured for transactions: sampled transactions have complete context recorded, and include spans;
+non-sampled transactions have limited context and no spans.
+
+For non-sampled transactions set the transaction attributes `sampled: false` and `sample_rate: 0`, and omit `context`.
```

Comment:
if `sample` = `false` and `sample_rate` is missing, the server will default the rate to 1, all while sampling being false, which doesn't make much sense. This opens the possibility for bugs. Am I missing something? Is there a way to be more semantically coherent?

Reply:
I'm not sure I get it. Why would `sample_rate` be missing? If `sampled: false` the `sample_rate` must be set to 0. It's just the transaction `context` that's being omitted. The `sample_rate` field is not part of `context`.

Reply:
I mean that is an optional field, so technically can be missing. I guess the "If sampled: false the sample_rate must be set to 0 because." kind of logic can slip quite easily (particularly thinking about future agents)

Reply:
> If sampled: false the sample_rate must be set to 0

I think this bit is more confusing now that we're sending `sample_rate` and not `sample_weight`...

> if sample = false and sample_rate is missing, the server will default the rate to 1, all while sampling being false, which doesn't make much sense. This opens the possibility for bugs. Am I missing something? Is there a way to be more semantically coherent?

The reason to have the server default the rate to 1 is to preserve backwards compatibility when calculating transaction histogram metrics. At the moment the server effectively assumes a rate of 1 for all transactions (excluding Jaeger). That is, sampled and non-sampled transactions are each counted as 1 in the histogram updates.

When the server sees `sample_rate: 0` it should treat that value specially and ignore the transaction when updating duration histogram metrics.

It is a little bit subtle. I'm open to alternative suggestions.

Reply:
Thanks @axw for explanations. 
What would be the implications of breaking backwards compatibility? 
Eg. what would happen if we default `sample_rate` to 0 when `sampled` is `false`, or 1 otherwise?

Reply:
If we defaulted to 0 for `sampled: false`, and 1 for `sampled: true`: for old agents that don't _ever_ send `sample_rate`, the transaction counts/rates would be under counted in histogram metrics. e.g.

Sampling rate is set to 0.5

Old agent:
- First transaction is sampled - transaction event sent with `sampled: true`
- Second transaction is unsampled - transaction event sent with `sampled: false`
- Server would count the first (default sample_rate=1), but not the second (default sample_rate=0) in histogram aggregation.

New agent:
- First transaction is sampled - transaction event sent with `sampled: true`, `sample_rate: 0.5`
- Second transaction is unsampled - transaction event sent with `sampled: false`, `sample_rate: 0`
- Server would count the first as 2 transactions (1/0.5), and would not count the second at all (sample_rate=0) in histogram aggregation.

Reply:
Ok :+1:  Maybe document the case than when all transactions are unsampled, none will count.

Reply:
@jalvz I've been wondering about this. I'm not sure if we should handle it or not.

@felixbarny in a meeting yesterday, @jalvz asked how we should deal with a user specifying a sampling rate of 0. In that case (with the current spec) the agent sets `sample_rate: 0` on every transaction, and the server will never update histogram metrics. The Metrics-based UI would then differ from our current behaviour, aggregating all (sampled and unsampled) transaction docs.

We could address that by having _another_ special case: if the user-specified sampling rate is 0, then for non-sampled transactions set `sample_rate: 1`. That would treat all (100% unsampled) transactions equally for the purpose of updating histograms.

But, if we want to stop sending unsampled transaction events to the server completely (like Jaeger, OpenTelemetry, ...), we can't do anything similar. In that case we would have to capture histograms in the agents.

Any thoughts on how we should proceed?

Reply:
I don't think we should special-case `0`. We have a similar issue when setting a really low sampling rate. The lower the rate, the less accurate the metrics will be. 

One special treatment we might want to consider, though, is to ensure that the minimum reported `sample_rate` is `0.001` if the `transaction_sampling_rate` is > 0. This avoids converting `0.0001` to `0`.

We're also not just calculating transaction metrics but also span metrics. We don't send any non-sampled spans. I assume the service map edge metrics will be based on the span metrics? Not sure about that. Might be a good addition to the spec. We should only have inconsistent ways of calculating metrics for transactions and spans if there's a very good reason. 

> But, if we want to stop sending unsampled transaction events to the server completely (like Jaeger, OpenTelemetry, ...), we can't do anything similar. In that case we would have to capture histograms in the agents.

Exactly, we shouldn't assume that we're sending up non-sampled transactions in the future. Every feature that relies on it avoids us from getting rid of that. If we want more accuracy for lower sampling rates, we have to track metrics on the agent and tell APM Server to omit metric tracking by setting a flag on the metadata.

Reply:
> The lower the rate, the less accurate the metrics will be

I don't think this is exactly the case. If sampling rate is 0, then the histogram will be missing / totally off, not simply "less accurate". IIUC.

I agree that *another* special case is not ideal. I don't know how complicated would be in the server to detect the case where all transactions in a trace have sampling rate 0, and then act accordingly. Probably *very* fiddly. 

> we would have to capture histograms in the agents

Why not go directly down that route? Or have the agents send the minimal aggregated data needed to a separate endpoint in apm-server, and have the server then to assemble histograms. I expect that could be done with fairly low overhead.

Reply:
@felixbarny 

> One special treatment we might want to consider, though, is to ensure that the minimum reported sample_rate is 0.001 if the transaction_sampling_rate is > 0. This avoids converting 0.0001 to 0.

That sounds fine, but then I think we should also ensure the agents round the sampling rate in the first place. i.e. state in docs that the precision is say 0.001 (sample 1 in 1000), and any value smaller than that would be rounded. If it would round to 0, the agents/UI should emit a warning that nothing will be sampled, and there will be no trace data or transaction metrics collected.

> We're also not just calculating transaction metrics but also span metrics. We don't send any non-sampled spans. I assume the service map edge metrics will be based on the span metrics? Not sure about that. Might be a good addition to the spec. We should only have inconsistent ways of calculating metrics for transactions and spans if there's a very good reason.

Great point. Yes, edge metrics == span metrics. I'll update the spec to describe how the `sample_rate` will be used by the server, including both transaction histograms and span/edge metrics.

@jalvz 

> I don't think this is exactly the case. If sampling rate is 0, then the histogram will be missing / totally off, not simply "less accurate". IIUC.

That's the ultimate inaccuracy isn't it? :)

We're intentionally trading off metrics accuracy here: the statistics will be based off sampled data only. When the Metrics-based UI is introduced we should be clear about that. Instead of setting sampling rate to 0 and expecting accurate metrics, users will have to set it to 0.1, 0.01 or 0.001 (depending on transaction rate).

Later we can consider improving accuracy by measuring transaction metrics at the agent. In that case metrics would not be based on sampling, and sample size/rate would not impact the metrics.

> I agree that another special case is not ideal. I don't know how complicated would be in the server to detect the case where all transactions in a trace have sampling rate 0, and then act accordingly. Probably very fiddly.

It's not impossible, but it is impractical. Not all events in the trace will necessarily go to the same server.

>> we would have to capture histograms in the agents
>
> Why not go directly down that route? Or have the agents send the minimal aggregated data needed to a separate endpoint in apm-server, and have the server then to assemble histograms. I expect that could be done with fairly low overhead.

Not sure what you mean by the "Or" there. The agents either have to send histograms (minimal aggregated data), or they have to send individual events (what we're doing today) for the server to create histograms. We want to move away from what we're doing today because it is expensive.

Sending individual (unsampled and sampled) transactions is also prone to incidental inaccuracy at high transaction rates, as it requires all events to make their way through to the server to calculate valid statistics. Basing the statistics off sampled data will alleviate this problem by intentionally only measuring sampled data.

I think we might want to implement agent-side aggregation _eventually_, but it's considerably more work for agents to implement, and we still need to implement the aggregations in the server for Jaeger and OpenTelemetry anyway.

Reply:
I've attempted to clarify all of the above in the renamed section: "Effect on metrics". PTAL

@felixbarny I didn't update it to say we should round sampling rate to 0.001. As above I think we should do that earlier (in configuration), and not just when recording the value in `tracestate`. Otherwise the metrics will be inconsistent with reality.

Reply:
Are you saying we should disallow sampling rates lower than 0.001?

Reply:
Opened https://github.com/elastic/apm/issues/322 to follow up on that

Reply:
> Later we can consider improving accuracy by measuring transaction metrics at the agent.

Cool, that is what I tried to say (not based on sampling) :+1: